### PR TITLE
Set Default `eps` Value To `jnp.finfo(float).eps` And Test

### PR DESCRIPTION
**Pull request by AFg6K7h4fhy2**
Previously the defaults for the `eps` value in `dist.Poisson(rate=mu + self.eps)` and `fn=dist.NegativeBinomial2(mean=mu + self.eps, concentration=concentration)` were set to `1e-8` and `1e-10` respectively, but instead can be set to the machine float minimum (see [here](https://numpy.org/doc/stable/reference/generated/numpy.finfo.html)) via `jnp.finfo(float).eps`. This needs to be tested, however, to make sure the distribution implementations of Poisson and NegativeBinomial still function properly.

### CODE REVIEW

**Comment by AFg6K7h4fhy2** [comment]
Would appreciate some thoughts on tests here. I am thinking testing new default, testing previous values, and testing a "large" value, where "testing" means "sampling does not through an error" and "samples are equal to jnp.ceil(samples)".

**Comment by damonbayer** [comment]
@AFg6K7h4fhy2 Testing should demonstrate that the naive case (`eps = 0`) fails in cases where using the new default `eps` succeeds.

File: model/src/test/test_rtperiodicdiff.py
Comment (damonbayer) [inline model/src/test/test_rtperiodicdiff.py:1]: This should probably be a separate PR. I'm also confused about why this fixes the tests.

Comment (AFg6K7h4fhy2) [reply model/src/test/test_rtperiodicdiff.py:1]: I am also confused why this fixes the tests as well. I think something is wrong with caching or awry with memory in `test_rtperiodicdiff.py`. I changed the `rng_seed` values in the PR for #307 to those in this PR and it failed. When I changed them to new values, the test passed. This ought to be investigated further.

For the separate PR point, I can delete the modified test in `test_observation_poisson.py` in this PR if you'd like. Or was the separate PR for investigating the situation in the previous paragraph?

Comment (damonbayer) [reply model/src/test/test_rtperiodicdiff.py:1]: Let's get the tests passing in a separate PR ASAP. Can make a separate issue to investigate why they weren't passing before.

Comment (damonbayer) [reply model/src/test/test_rtperiodicdiff.py:1]: Resolved in https://github.com/CDCgov/multisignal-epi-inference/pull/327. You can remove the changes here.

**Comment by damonbayer** [comment]
This problem is probably better addressed via jax.numpy.clip  https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.clip.html

**Comment by AFg6K7h4fhy2** [comment]
> This problem is probably better addressed via jax.numpy.clip https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.clip.html

Not quite sure I get this. Rather than have a machine-minimum, we want to clip `esp` to be between `jnp.finfo(float).eps` and some other value (upper bound)?

**Comment by damonbayer** [comment]
> Not quite sure I get this. Rather than have a machine-minimum, we want to clip `esp` to be between `jnp.finfo(float).eps` and some other value (upper bound)?

The problem we are trying to avoid is that the mean argument to `sample`  maybe be too small. `clip`ping the value to be between (whatever the smallest valid value is) and infinity makes the `mean` argument valid while keeping it as close to the input `mean` value as possible.


With this approach, you may want to error if the `mean` argument is too negative.

**Comment by AFg6K7h4fhy2** [comment]
Came back to this today. Can you take a look @damonbayer ?

File: model/src/pyrenew/observation/poisson.py [hunk 1]
```diff
@@ -67,9 +68,13 @@ def sample(
         tuple
         """
 
+        clipped_mu = jnp.clip(
+            mu + self.eps, min=jnp.finfo(float).eps, max=jnp.inf
```
Comment (damonbayer) [inline model/src/pyrenew/observation/poisson.py:72]: We should just use clips, no need to add eps.

File: model/src/test/test_observation_negativebinom.py
Comment (damonbayer) [inline model/src/test/test_observation_negativebinom.py:1]: There should be a test that demonstrates the value of using `clip` or `eps`. We should to show a case where the naive method fails, but our approach succeeds.

Comment (damonbayer) [reply model/src/test/test_observation_negativebinom.py:1]: The `eps` stuff was added before I joined the project, so I'm not completely sure what specific situation this is meant to remedy.

Comment (AFg6K7h4fhy2) [reply model/src/test/test_observation_negativebinom.py:1]: Going to tag @dylanhmorris here before I remove `eps`.

Comment (dylanhmorris) [reply model/src/test/test_observation_negativebinom.py:1]: If we can't make a reprex of failure it may be that things now work and we can remove `clip`-ing / `eps`-addition entirely. In the past allowing for literal 0 negative binomial / poisson means could break things, but maybe this has now been addressed.

Comment (AFg6K7h4fhy2) [reply model/src/test/test_observation_negativebinom.py:1]: I think DHM's remark — — holds here.

The following test fails because the naive case here does not fail.

Note that this is what the clipping looks like in `sample` function of NegativeBinomialObservation:

```python
clipped_mu = jnp.clip(mu + self.eps, min=self.eps, max=jnp.inf)
```

Test that does not pass currently: 

```python
def test_negativebinom_naive_vs_clipped():
    """
    Test that demonstrates the failure of the naive method (eps=0) 
    and the success of the method with clipping or small eps.
    """

  
    concentration_rv = DeterministicVariable(name="concentration", value=10)

    negb_naive = NegativeBinomialObservation(
        "negbinom_rv_naive",
        concentration_rv=concentration_rv,
        eps=0.0
    )

    negb_clipped = NegativeBinomialObservation(
        "negbinom_rv_clipped",
        concentration_rv=concentration_rv,
        eps=jnp.finfo(float).eps
    )

    naive_failed = False
    try:
        with numpyro.handlers.seed(rng_seed=223):
            negb_naive_sample = negb_naive(mu=0.0)
    except Exception as e:
        naive_failed = True
        print(f"Naive method failed as expected: {e}")
    assert naive_failed, "Naive method did not fail as expected."

    clipped_failed = False
    try:
        with numpyro.handlers.seed(rng_seed=223):
            negb_clipped_sample = negb_clipped(mu=0.0)
    except Exception as e:
        clipped_failed = True
        print(f"Clipped method failed unexpectedly: {e}")
    assert not clipped_failed, "Clipped method failed unexpectedly."

    assert isinstance(negb_clipped_sample, tuple)
    assert isinstance(negb_clipped_sample[0].value, ArrayLike)

    mean_sample_value = jnp.mean(negb_clipped_sample[0].value)
    assert mean_sample_value >= jnp.finfo(float).eps, "Clipped sample mean is unexpectedly low."
```

Comment (AFg6K7h4fhy2) [reply model/src/test/test_observation_negativebinom.py:1]: Further thoughts here @dylanhmorris ?

Comment (damonbayer) [reply model/src/test/test_observation_negativebinom.py:1]: I think I would be willing to close this PR and accept a new one where we just set `eps=0` as the default. Having the option of `eps` seems like it could be useful in the future? If we find a use case for it, we can re-visit this issue.

Comment (dylanhmorris) [reply model/src/test/test_observation_negativebinom.py:1]: Sounds good to me.

Comment (AFg6K7h4fhy2) [reply model/src/test/test_observation_negativebinom.py:1]: My thinking is similar if you are also thinking of keeping the clipping in place. "Doesnt hurt might be useful" is where I am at. 

>  If we find a use case for it, we can re-visit this

+1

Comment (damonbayer) [reply model/src/test/test_observation_negativebinom.py:1]: For now, I prefer no clipping, and eps of 0.
